include/kernel.h

```diff
@@ -2469,930 +2469,1577 @@ __syscall int k_stack_pop(struct k_stack *stack, stack_data_t *data,
+ *
+ * @param delay the time to wait before submitting the work item.  If @c
+ * K_NO_WAIT this is equivalent to k_work_submit_to_queue().
+ *
+ * @return as with k_work_schedule_for_queue().
  */
+static inline int k_work_schedule(struct k_work_delayable *dwork,
+				   k_timeout_t delay)
+{
+	return k_work_schedule_for_queue(&k_sys_work_q, dwork, delay);
+}
 
-/**
- * @defgroup semaphore_apis Semaphore APIs
- * @ingroup kernel_apis
- * @{
+/** @brief Reschedule a work item to a queue after a delay.
+ *
+ * Unlike k_work_schedule_for_queue() this function can change the deadline of
+ * a scheduled work item, and will schedule a work item that isn't idle
+ * (e.g. is submitted or running).  This function does not affect ("unsubmit")
+ * a work item that has been submitted to a queue.
+ *
+ * @note Safe to invoke from ISRs.
+ *
+ * @param queue the queue on which the work item should be submitted after the
+ * delay.
+ *
+ * @param dwork pointer to the delayable work item.
+ *
+ * @param delay the time to wait before submitting the work item.  If @c
+ * K_NO_WAIT this is equivalent to k_work_submit_to_queue() after canceling
+ * any previous scheduled submission.
+ *
+ * @note If delay is @c K_NO_WAIT ("no delay") the return values are as with
+ * k_work_submit_to_queue().
+ *
+ * @retval 0 if delay is @c K_NO_WAIT and work was already on a queue
+ * @retval 1 if
+ * * delay is @c K_NO_WAIT and work was not submitted but has now been queued
+ *   to @p queue; or
+ * * delay not @c K_NO_WAIT and work has been scheduled
+ * @retval 2 if delay is @c K_NO_WAIT and work was running and has been queued
+ * to the queue that was running it
  */
+int k_work_reschedule_for_queue(struct k_work_q *queue,
+				 struct k_work_delayable *dwork,
+				 k_timeout_t delay);
 
-/**
- * @brief Initialize a semaphore.
+/** @brief Reschedule a work item to the system work queue after a
+ * delay.
  *
- * This routine initializes a semaphore object, prior to its first use.
+ * This is a thin wrapper around k_work_reschedule_for_queue(), with all the
+ * API characteristcs of that function.
  *
- * @param sem Address of the semaphore.
- * @param initial_count Initial semaphore count.
- * @param limit Maximum permitted semaphore count.
+ * @param dwork pointer to the delayable work item.
  *
- * @retval 0 Semaphore created successfully
- * @retval -EINVAL Invalid values
+ * @param delay the time to wait before submitting the work item.
  *
+ * @return as with k_work_reschedule_for_queue().
  */
-__syscall int k_sem_init(struct k_sem *sem, unsigned int initial_count,
-			  unsigned int limit);
+static inline int k_work_reschedule(struct k_work_delayable *dwork,
+				     k_timeout_t delay)
+{
+	return k_work_reschedule_for_queue(&k_sys_work_q, dwork, delay);
+}
 
-/**
- * @brief Take a semaphore.
+/** @brief Flush delayable work.
  *
- * This routine takes @a sem.
+ * If the work is scheduled, it is immediately submitted.  Then the caller
```

Comment:
i have some doubts regarding this functionality. Work was delayed for some reason, now when we want to sync until it is finished we are also getting it rescheduled to immediate execution. It seems to me that it would be good to have an option to choose whether i want to reschedule or wait until it is executed at scheduled time.

Reply:
Perhaps, but this is the behavior of the corresponding Linux API.  This behaves "as expected" when the work item is both scheduled and submitted: on flush the work is completely done, not potentially completing multiple times.

The infrastructure to handle "wait until it's scheduled, and then wait until it's done" could get complicated, and it's not obvious that's useful.